File: medium/search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        l = 0 
        r = len(matrix)-1
        while l <= r:
            m = l + (r - l) // 2
            if matrix[m][0] == target:
                return True
            if matrix[m][len(matrix[m])-1] == target:
                return True
            if matrix[m][0] < target and matrix[m][len(matrix[m])-1] > target:
                k = 0
                e = len(matrix[m])-1
                while k <= e:
                    n = k + (e - k) // 2
                    if matrix[m][k] == target:
                        return True
                    if matrix[m][e] == target:
                        return True
                    if matrix[m][n] == target:
                        return True
                    if matrix[m][n] > target:
                        e = n - 1
                    if matrix[m][n] < target:
                        k = n + 1
                return False
            if matrix[m][0] > target:
                r = m - 1
                continue
            if matrix[m][0] < target:
                l = m + 1
                continue
            return False

File: medium/search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int t = 0;
        int b = matrix.size()-1;
        while (t <= b) {
            int m = t + (b-t) / 2;
            if (matrix[m][0] == target) {
                return true;
            }
            if (matrix[m][matrix[m].size()-1] == target) {
                return true;
            }
            if (matrix[m][0] < target && matrix[m][matrix[m].size()-1] > target) {
                int l = 0;
                int r = matrix[m].size();
                while (l <= r) {
                    int n = l + (r-l) / 2;
                    if (matrix[m][n] == target) {
                        return true;
                    }
                    if (matrix[m][n] > target) {
                        r = n - 1;
                    }
                    if (matrix[m][n] < target) {
                        l = n + 1;
                    }
                }
                return false;
            }
            if (matrix[m][0] < target && matrix[m][matrix[m].size()-1] < target) {
                t = m + 1;
            }
            if (matrix[m][0] > target && matrix[m][matrix[m].size()-1] > target) {
                b = m - 1;
            }
        }
        return false;
    }
};
